Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,258 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json;
+
+using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Http;
+using Google.Apis.Json;
+using Google.Apis.Logging;
+using Google.Apis.Util;
+
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>
+    /// Provides the Application Default Credential from the environment. 
+    /// An instance of this class represents the per-process state used to get and cache 
+    /// the credential and allows overriding the state and environment for testing purposes.
+    /// </summary>
+    internal class DefaultCredentialProvider
+    {
+        private static readonly ILogger Logger = ApplicationContext.Logger.ForType<DefaultCredentialProvider>();
+
+        /// <summary>Environment variable override which stores the default application credentials file path.</summary>
+        public const string CredentialEnvironmentVariable = "GOOGLE_APPLICATION_CREDENTIALS";
+
+        /// <summary>Well known file which stores the default application credentials.</summary>
+        private const string WellKnownCredentialsFile = "application_default_credentials.json";
+
+        /// <summary>Environment variable which contains the Application Data settings.</summary>
+        private const string AppdataEnvironmentVariable = "APPDATA";
+
+        /// <summary>Environment variable which contains the location of home directory on UNIX systems.</summary>
+        private const string HomeEnvironmentVariable = "HOME";
+
+        /// <summary>GCloud configuration directory in Windows, relative to %APPDATA%.</summary>
+        private const string CloudSDKConfigDirectoryWindows = "gcloud";
+
+        /// <summary>Help link to the application default credentials feature.</summary>
+        private const string HelpPermalink = "https://developers.google.com/accounts/docs/application-default-credentials";
+
+        /// <summary>GCloud configuration directory on Linux/Mac, relative to $HOME.</summary>
+        private static readonly string CloudSDKConfigDirectoryUnix = Path.Combine(".config", "gcloud");
+
+        /// <summary>Caches result from first call to <c>GetApplicationDefaultCredentialAsync</c> </summary>
+        private readonly Lazy<Task<GoogleCredential>> cachedCredentialTask;
+
+        public DefaultCredentialProvider()
+        {
+            this.cachedCredentialTask = new Lazy<Task<GoogleCredential>>(CreateDefaultCredentialAsync);
```

Comment:
Can we remove "this" I don't think that you need it.
cachedCredentialTask = new Lazy<Task<GoogleCredential>>(CreateDefaultCredentialAsync);